### PR TITLE
[SPARK-15279][SQL] Catch conflicting SerDe when creating table

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?

The user may do something like:

```
CREATE TABLE my_tab ROW FORMAT SERDE 'anything' STORED AS PARQUET
CREATE TABLE my_tab ROW FORMAT SERDE 'anything' STORED AS ... SERDE 'myserde'
CREATE TABLE my_tab ROW FORMAT DELIMITED ... STORED AS ORC
CREATE TABLE my_tab ROW FORMAT DELIMITED ... STORED AS ... SERDE 'myserde'
```

None of these should be allowed because the SerDe's conflict. As of this patch:
- `ROW FORMAT DELIMITED` is only compatible with `TEXTFILE`
- `ROW FORMAT SERDE` is only compatible with `TEXTFILE`, `RCFILE` and `SEQUENCEFILE`
## How was this patch tested?

New tests in `DDLCommandSuite`.

### CODE REVIEW

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
@yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 1]
```diff
@@ -941,6 +944,43 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+    if (rowFormatCtx == null || createFileFormatCtx == null) {
+      return
+    }
+    val cff = (0 until createFileFormatCtx.getChildCount)
+      .map { i => createFileFormatCtx.getChild(i).getText }
+      .mkString(" ")
+    (rowFormatCtx, createFileFormatCtx.fileFormat) match {
+      case (_, ffTable: TableFileFormatContext) =>
+        if (visitTableFileFormat(ffTable).serde.isDefined) {
+          throw operationNotAllowed(s"ROW FORMAT is not compatible with $cff", parentCtx)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Maybe we can be explicit that `SERDE` is not allowed for this case? btw, should we remove `SERDE` part from `INPUTFORMAT inFmt=STRING OUTPUTFORMAT outFmt=STRING (SERDE serdeCls=STRING)?` in `SqlBase.g4`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 2]
```diff
@@ -941,6 +944,43 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+        }
+      case (rfSerde: RowFormatSerdeContext, ffGeneric: GenericFileFormatContext) =>
+        ffGeneric.identifier.getText.toLowerCase match {
+          case ("sequencefile" | "textfile" | "rcfile") => // OK
+          case _ => throw operationNotAllowed(
+            s"ROW FORMAT SERDE is not compatible with $cff", parentCtx)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Maybe we can say something like `Format ${ffGeneric.identifier} does not a user-specific SerDe`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 3]
```diff
@@ -941,6 +944,43 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+        }
+      case (rfDelimited: RowFormatDelimitedContext, ffGeneric: GenericFileFormatContext) =>
+        ffGeneric.identifier.getText.toLowerCase match {
+          case "textfile" => // OK
+          case _ => throw operationNotAllowed(
+            s"ROW FORMAT SERDE is not compatible with $cff", parentCtx)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: `ROW FORMAT DELIMITED` is only allowed to use with `textfile`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 4]
```diff
@@ -941,6 +944,43 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
...
+        }
+      case (rf, ff) =>
+        // should never happen
+        throw operationNotAllowed(s"Unexpected combination of ROW FORMAT and $cff", parentCtx)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Should we also print out `rf`? Is `ff` the same as `cff`?

(Maybe it is good to add a few examples at here?)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 5]
```diff
@@ -941,6 +944,43 @@ class SparkSqlAstBuilder(conf: SQLConf) extends AstBuilder {
   }
 
   /**
+   * Throw a [[ParseException]] if the user specified incompatible SerDes through ROW FORMAT
+   * and STORED AS.
+   */
+  private def validateRowFormatFileFormat(
+      rowFormatCtx: RowFormatContext,
+      createFileFormatCtx: CreateFileFormatContext,
+      parentCtx: ParserRuleContext): Unit = {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: (Maybe it is good to add a few examples to this?)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 3 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: isn't that what we talked about?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 4 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: cff is the nicely formatted string

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 1 above)
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: yeah, I like that actually! Hive doesn't support STORED AS serde too: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 3 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: oh, sorry. I was thinking about what the error message should say.

Yea, the semantic is what we talked about.

**Comment by andrewor14** [comment]
All comments have been addressed

**Comment by yhuai** [comment]
Thanks! LGTM

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
m2.0

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 1 above)
Comment (StefanXiepj) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala]: Why don't we support this? What's the problem? Can you give me an example? thx

**Pull request closed without merging**